enhancements/cluster-logging/cluster-logging-log-forwarding.md

```diff
@@ -0,0 +1,310 @@
+
+
+
+### Goals
+The specific goals of this proposal are:
+
+* Provide an API that identifies log sources and their desired destination
+* Send logs to an Elasticsearch cluster not managed by the cluster logging infrastructure
+* Send logs to a fluentd instance not managed by the cluster logging infrastructure via fluentd's secureforward protocol
+* Allow deployment of log forwarding without deploying the entirety of the cluster logging infrastructure (e.g. Kibana, Elasticsearch)
+* Support TLS between the collector and destination if so configured
+* Specifically only forward "application" and "infrastructure" logs
+* Re-use log forwarding to write logs to a cluster logging managed log aggregator (e.g. Elasticsearch)
+
+We will be successful when:
+
+* an administrator is able to deploy their own log aggregation service
+* specifies this service in log forwarding configuration
+* the service receives logs from a OKD cluster logging deployed collection agent  
+
+
+
+### Non-Goals
+
+Log forward intends to be opinionated about the source of log messages; it will explicitly define what constitutes "application logs" or "infrastructure logs". It will not allow configuration of additional sources without further design considerations and how those sources may impact the collecting agent. Log forwarding is not intended to provide a complex routing solution as one might achieve by using a custom collector configuration or a messaging solution (e.g. kafka).
+
+## Proposal
+
+Log forwarding will provide a declarative way to specify the destination for specific types of logs using a 'pipeline'.  A 'pipeline' defines simple routing for one source to one or more destinations. The source of logs are opinionated and well defined by cluster logging. The initial source types are as follows:
+
+* `logs.app` - Container logs generated by user applications running on the platform, excluding infrastructure container applications
+* `logs.infra` - Logs generated by both infrastructure components running on the platform and OKD nodes (e.g. journal logs).  "Infra" applications are defined as any pods which run in namespaces: `openshift*`, `kube*`, `default`.
+
+Future types may be:
+* `events.kubernetes` - Kubernetes events
+* `audit.container-engine` - Container engine audit events
+
+There are no assumptions regarding whether or not an endpoint is deployed on or off cluster.  Endpoints off-cluster may require adminstrators to perform additional actions in order for logs to be forwarded (e.g. secret creation, opening port) 
+Following is the list of supported endpoint types for this proposal:
+
+* Elasticsearch
+* Fluent secure-forward
+
+### User Stories
+
+#### As an OKD admin, I want to collect logs and send them to my own aggregation solution, for example Splunk
+
+This is a typical example of organizations that desires to re-use their existing enterprise log solution.  We will succeeded if we are able to write logs to their logging service.
```

Comment:
Splunk isn't listed above as a supported endpoint, so how will this be enabled?  via secure-forward to a secondary fluentd(installed/managed by the admin) that the customer then configures to send to splunk?

Reply:
Right - maybe this should be the user story for deploying only the collector with no Elasticsearch/Kibana - "As an OKD admin, I want to deploy only the OKD log collector, and have it ship logs off cluster to my own aggregation solution."